src/graph/InsertVertexExecutor.cpp

```diff
@@ -26,91 +25,128 @@ Status InsertVertexExecutor::prepare() {
+            if (schema == nullptr) {
+                return Status::Error("No schema found for `%s'", tagName->c_str());
+            }
+
+            auto props = item->properties();
+            // Now default value is unsupported, props should equal to schema's fields
+            if (schema->getNumFields() != props.size()) {
+                LOG(ERROR) << "props number " << props.size()
+                           << ", schema field number " << schema->getNumFields();
+                return Status::Error("Wrong number of props");
+            }
+
+            tagIds_.emplace_back(tagId);
+            schemas_.emplace_back(schema);
+            tagProps_.emplace_back(props);
+
+            // Check field name
+            auto checkStatus = checkFieldName(schema, props);
+            if (!checkStatus.ok()) {
+                return checkStatus;
+            }
         }
     } while (false);
 
     return status;
 }
 
 
-void InsertVertexExecutor::execute() {
-    using storage::cpp2::Vertex;
-    using storage::cpp2::Tag;
-
-    std::vector<Vertex> vertices(rows_.size());
+StatusOr<std::vector<storage::cpp2::Vertex>> InsertVertexExecutor::prepareVertices() {
+    std::vector<storage::cpp2::Vertex> vertices(rows_.size());
     for (auto i = 0u; i < rows_.size(); i++) {
         auto *row = rows_[i];
         auto id = row->id();
         auto expressions = row->values();
-        std::vector<VariantType> values;
 
+        std::vector<VariantType> values;
         values.reserve(expressions.size());
         for (auto *expr : expressions) {
             values.emplace_back(expr->eval());
         }
 
-        auto &vertex = vertices[i];
-        std::vector<Tag> tags(1);
-        auto &tag = tags[0];
-        RowWriter writer(schema_);
-
-        for (auto &value : values) {
-            switch (value.which()) {
-                case 0:
-                    writer << boost::get<int64_t>(value);
-                    break;
-                case 1:
-                    writer << boost::get<double>(value);
-                    break;
-                case 2:
-                    writer << boost::get<bool>(value);
-                    break;
-                case 3:
-                    writer << boost::get<std::string>(value);
-                    break;
-                default:
-                    LOG(FATAL) << "Unknown value type: " << static_cast<uint32_t>(value.which());
+        storage::cpp2::Vertex vertex;
+        std::vector<storage::cpp2::Tag> tags(tagIds_.size());
+
+        auto valuePos = 0u;
+        for (auto index = 0u; index < tagIds_.size(); index++) {
+            auto &tag = tags[index];
+            auto tagId = tagIds_[index];
+            auto propSize = tagProps_[index].size();
+            auto schema = schemas_[index];
+
+            // props's number should equal to value's number
+            if ((valuePos + propSize) > values.size()) {
```

Comment:
the comment is confusing with code.
if ">" is not precise enough. use == compare instead.

Reply:
It doesn't. Because multi tags' props are saved in the same set of values. It handles one by one tag, so it judges the sum of tag’s props every time